### DISCUSSION TITLE
pytest fixture teardown in not used test

### DISCUSSION THREAD

**Discussion by dinartal** [pytest-dev/pytest#9874] [Q&A]
Hello, i want to group some tests and execute them together. Also want some code be done before and after this tests.
So i created fixture and use it in some tests but teardown code after yield execute after whole session, after test that not use this fixture. For example:

conftest.py:
@pytest.fixture(scope='session', autouse=False)
def example_f():
    logging.info("BEFORE")
    yield
    logging.info("AFTER")

test_1.py:
import logging
def test_1(example_f):
    logging.info("TEST 1")

test_2.py:
import logging
def test_2():
    logging.info("TEST 2")

And here is execution output:
2022-04-21 19:04:59.050 [    INFO] BEFORE (conftest.py:51)
2022-04-21 19:04:59.051 [    INFO] TEST 1 (Test_1.py:3)
test_1 ✓                                                                                                                50% █████     
2022-04-21 19:04:59.058 [    INFO] TEST 2 (Test_2.py:3)
2022-04-21 19:04:59.058 [    INFO] AFTER (conftest.py:53)
test_2 ✓ 

But i want next behaviour:
2022-04-21 19:04:59.050 [    INFO] BEFORE (conftest.py:51)
2022-04-21 19:04:59.051 [    INFO] TEST 1 (Test_1.py:3)
2022-04-21 19:04:59.058 [    INFO] AFTER (conftest.py:53)
2022-04-21 19:04:59.058 [    INFO] TEST 2 (Test_2.py:3)

**Comment by dinartal** [comment]
[](https://docs.pytest.org/en/latest/how-to/fixtures.html#fixture-scopes)Fixture scopes
Fixtures are created when first requested by a test, and are destroyed based on their scope:

function: the default scope, the fixture is destroyed at the end of the test.

class: the fixture is destroyed during teardown of the last test in the class.

module: the fixture is destroyed during teardown of the last test in the module.

package: the fixture is destroyed during teardown of the last test in the package.

session: the fixture is destroyed at the end of the test session.

**Comment by dinartal** [comment]
Do do it i need to put tests in one file..

**Reply by RonnyPfannschmidt** [reply]
There are some issues about more eager teardown of no longer needed fixtures, but so far implementation eluded us

**Reply by The-Compiler** [reply]
I believe this is working as designed. Consider a test using `request.getfixturevalue()` with such a session-scoped fixture - what would happen? Suddenly having two instances of a fixture with `session` caching scope sounds like a recipe for trouble.
